akka-stream-tests/src/test/scala/akka/stream/impl/fusing/AsyncCallbackSpec.scala

```diff
@@ -0,0 +1,228 @@
+        val msg = "whatever" + n
+        val feedbackF = callback.invokeWithFeedback(msg)
+        probe.expectMsg(msg)
+        feedbackF.futureValue should ===(Done)
+      }
+      in.sendComplete()
+      out.expectComplete()
+
+      probe.expectMsg(Stopped)
+    }
+
+    "fail the feedback future if stage is stopped" in {
+      val probe = TestProbe()
+      val callbackF = Source.empty
+        .viaMat(new AsyncCallbackGraphStage(probe.ref))(Keep.right)
+        .to(Sink.ignore)
+        .run()
+
+      probe.expectMsg(Started)
+      probe.expectMsg(Stopped)
+
+      val feedbakF = callbackF.futureValue.invokeWithFeedback("whatever")
+      feedbakF.failed.futureValue shouldBe a[StreamDetachedException]
+    }
+
+    "invoke early" in {
+      val probe = TestProbe()
+      val in = TestPublisher.probe[Int]()
+      val callbackF = Source.fromPublisher(in)
+        .viaMat(new AsyncCallbackGraphStage(
+          probe.ref,
+          Some(asyncCb ⇒ asyncCb.invoke("early"))
+        ))(Keep.right)
+        .to(Sink.ignore)
+        .run()
+
+      // and deliver in order
+      val callback = callbackF.futureValue
+      callback.invoke("later")
+
+      probe.expectMsg(Started)
+      probe.expectMsg("early")
+      probe.expectMsg("later")
+
+      in.sendComplete()
+      probe.expectMsg(Stopped)
+
+    }
+
+    "invoke with feedback early" in {
+      val probe = TestProbe()
+      val earlyFeedback = Promise[Done]()
+      val in = TestPublisher.probe[Int]()
+      val callbackF = Source.fromPublisher(in)
+        .viaMat(new AsyncCallbackGraphStage(
+          probe.ref,
+          Some(asyncCb ⇒ earlyFeedback.completeWith(asyncCb.invokeWithFeedback("early")))
+        ))(Keep.right)
+        .to(Sink.ignore)
+        .run()
+
+      // and deliver in order
+      val callback = callbackF.futureValue
+      val laterFeedbackF = callback.invokeWithFeedback("later")
+
+      probe.expectMsg(Started)
+      probe.expectMsg("early")
+      earlyFeedback.future.futureValue should ===(Done)
+
+      probe.expectMsg("later")
+      laterFeedbackF.futureValue should ===(Done)
+
+      in.sendComplete()
+      probe.expectMsg(Stopped)
+    }
+
+    "accept concurrent input" in {
+      val probe = TestProbe()
+      val in = TestPublisher.probe[Int]()
+      val callbackF = Source.fromPublisher(in)
+        .viaMat(new AsyncCallbackGraphStage(probe.ref))(Keep.right)
+        .to(Sink.ignore)
+        .run()
+
+      val callback = callbackF.futureValue
+
+      import system.dispatcher
+      val feedbacks = (1 to 100).map { n ⇒
+        Future {
+          callback.invokeWithFeedback(n.toString)
+        }.flatMap(d ⇒ d)
+      }
+
+      probe.expectMsg(Started)
+      Future.sequence(feedbacks).futureValue should have size (100)
+      (1 to 100).map(_ ⇒ probe.expectMsgType[String]).toSet should have size (100)
+
+      in.sendComplete()
+      probe.expectMsg(Stopped)
+    }
+
+    "fail the feedback if the invoke fails" in {
```

Comment:
Not 100% about this behavior but it felt right when I added it :[

Reply:
Seems logical to me, what's the case against failing it like that?